slotsmanager.c

```diff
@@ -127,46 +130,90 @@ slots_gc(void)
      * link lists first. likely worth it, since GC happens only in domain decompose
      * and snapshot IO, both take far more time than rebuilding the tree. */
     tree_invalid |= slots_gc_base();
-    tree_invalid |= slots_gc_slots();
+    tree_invalid |= slots_gc_slots(compact_slots);
 
     MPI_Allreduce(MPI_IN_PLACE, &tree_invalid, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
 
-    EISlotsAfterGC event = {0};
+    EISlotsAfterGC event = {{tree_invalid}};
 
     event_emit(&EventSlotsAfterGC, (EIBase*) &event);
```

Comment:
What if we add rebuild_activelist as an event handler?

Reply:
We could - I'd rather not have it in a whole handler of it's own. The point of adding it is to ensure that when the tree gets rebuilt, the active list also gets rebuilt, so we should have them in the same function. But adding it to the existing handler would introduce a cross-module dependency of forcetree.c on timestep.c, and I'm not sure this is worth it, since at the moment this is only done in one place. What do you think?

Reply:
Hmm, no, on second thoughts lets have it in its own handler.

Reply:
If it is in its own handler, then the active list will also be rebuilt after slot_gc in domain_exchange: so we would remove the rebuild_activelist call in the mainloop. But then we have a problem, which is that if the exchange does not generate garbage we would still need to rebuild the active list.

Also hiding something that is really a mainloop action in an event handler makes things harder to follow: working out what happens in event handlers is a bit tricky and over-using them can lead to crazy spaghetti code (parts of KDE in particular suffer from this). So I think on balance I prefer not to add it as an event handler.

Reply:
The better fix might be to move force_tree_rebuild back into the main loop and send the event handler back to being an error.

Reply:
I have done this - the event handler is now gone.

Reply:
Sure. The event handlers are to decouple interaction between modules -- the missing part is probably a diagram that charts who emits what and who listen to what -- in UI toolkits it's been too frequently abused as a way to avoid the overhead cost of subclassing -- I suppose that's what makes you dislike Events.

I agree that if the mainloop already depend on all modules then there is no need to use it. I thought it might be cleaner to keep all forcetree logic in forcetree and timebin logic in timebin -- where a very thin mainloop does almost nothing. So another approach is to say the event handler has an argument about number of baseslots moved. Then we can decide if to rebuild the tree in the force tree handler, and decide to rebuild the activelist in the timebin handler.

Reply:
I don't dislike events per se - I think the current post-fork event is an excellent use of them, because it does remove cross-module dependencies and it describes something that could conceptually happen at any time in the code, so maps well onto what I understand an event to be. 

I just don't like over-using them: I saw once a mainloop which was implemented by several events chaining themselves together, which was just really hard to follow. Because they are attached at runtime it can become hard to follow which event handlers are actually running and debugging becomes tricky. I fixed a bug in nepomuk once where an event was being disconnected under peculiar circumstances deep within some other handler - it took me forever to figure it out, because I didn't have a way to find the function disconnecting it.

So if the main loop depends on all modules then it should explicitly state that in the header files - hiding the dependency in an event is just confusing.

One good reason for events is optional extensibility - I can see several of our feedback modules being implemented that way eventually. For example, SPH_GRAD_RHO could be replaced with an event handler registered by the H2 star formation criterion.